README.md
# @coinset/ftx

Universal FTX API client

:children_crossing: This is not official

## Public API

A request for an entry point that does not require authentication.

### fetchMarket

Return type of market on FTX: spot, perpetual future, expiring future, and MOVE
contract. Examples for each type are `BTC/USD`, `BTC-PERP`, `BTC-0626`, and
`BTC-MOVE-1005`. For future that expired in 2019, prepend a `2019` to the date,
like so: `BTC-20190628` or `BTC-MOVE-20190923`.
[Docs](https://docs.ftx.com/?javascript#get-single-market)

example:

```ts
import { fetchMarket } from "https://deno.land/x/ftx@$VERSION/mod.ts";
await fetchMarket("BTC/USD");
```

parameter:

| name       | type     | required | description                                         |
| ---------- | -------- | -------- | --------------------------------------------------- |
| marketName | `string` | *        | e.g. `"BTC/USD"` for spot, `"BTC-PERP"` for futures |

returns:

```ts
type BaseMarketData = {
  quoteVolume24h: number;
  change1h: number;
  change24h: number;
  changeBod: number;
  highLeverageFeeExempt: boolean;
  minProvideSize: number;
  enabled: boolean;
  ask: number | null;
  bid: number | null;
  last: number | null;
  price: number | null;
  postOnly: boolean;
  priceIncrement: number;
  sizeIncrement: number;
  restricted: boolean;
  volumeUsd24h: number;
  tokenizedEquity?: boolean | undefined;
};

type MarketResponse = {
  success: true;
  result: (
    & BaseMarketData
    & ({
      name: `${string}/${string}`;
      baseCurrency: string;
      quoteCurrency: string;
      underlying: null;
      type: "spot";
    } | {
      name: string;
      baseCurrency: null;
      quoteCurrency: null;
      underlying: string;
      type: "future";
    })
  );
};
```

### fetchMarkets

This section covers all types of markets on FTX: spot, perpetual futures,
expiring futures, and MOVE contracts. Examples for each type are `BTC/USD`,
`BTC-PERP`, `BTC-0626`, and `BTC-MOVE-1005`. For futures that expired in 2019,
prepend a `2019` to the date, like so: `BTC-20190628` or `BTC-MOVE-20190923`.
[Docs](https://docs.ftx.com/?javascript#markets)

example:

```ts
import { fetchMarkets } from "https://deno.land/x/ftx@$VERSION/mod.ts";
await fetchMarkets();
```

returns:

```ts
type BaseMarketData = {
  quoteVolume24h: number;
  change1h: number;
  change24h: number;
  changeBod: number;
  highLeverageFeeExempt: boolean;
  minProvideSize: number;
  enabled: boolean;
  ask: number | null;
  bid: number | null;
  last: number | null;
  price: number | null;
  postOnly: boolean;
  priceIncrement: number;
  sizeIncrement: number;
  restricted: boolean;
  volumeUsd24h: number;
  tokenizedEquity?: boolean | undefined;
};

type MarketsResponse = {
  success: true;
  result: (
    & BaseMarketData
    & ({
      name: `${string}/${string}`;
      baseCurrency: string;
      quoteCurrency: string;
      underlying: null;
      type: "spot";
    } | {
      name: string;
      baseCurrency: null;
      quoteCurrency: null;
      underlying: string;
      type: "future";
    })
  )[];
};
```

### fetchOrderBook

Retrieve order book of market name.
[Docs](https://docs.ftx.com/?javascript#get-orderbook)

example:

```ts
import { fetchOrderBook } from "https://deno.land/x/ftx@$VERSION/mod.ts";
await fetchOrderBook("BTC/USD");
```

parameter:

| name       | type                 | required | description                                         |
| ---------- | -------------------- | -------- | --------------------------------------------------- |
| marketName | `string`             | *        | e.g. `"BTC/USD"` for spot, `"BTC-PERP"` for futures |
| options    | `{ depth?: number }` |          | order book options                                  |

returns:

```ts
type OrderBookResponse = {
  success: true;
  result: {
    asks: [number, number][];
    bids: [number, number][];
  };
};
```

### fetchTrades

Retrieve list of trade history.
[Docs](https://docs.ftx.com/?javascript#get-trades)

example:

```ts
import { fetchTrades } from "https://deno.land/x/ftx@$VERSION/mod.ts";
await fetchTrades("BTC/USD");
```

parameter:

| name       | type                                       | required | description                                         |
| ---------- | ------------------------------------------ | -------- | --------------------------------------------------- |
| marketName | `string`                                   | *        | e.g. `"BTC/USD"` for spot, `"BTC-PERP"` for futures |
| options    | `{ startTime?: number, endTime?: number }` |          | trades options                                      |

returns:

```ts
type TradesResponse = {
  success: true;
  result: {
    id: number;
    liquidation: boolean;
    price: number;
    side: "buy" | "sell";
    size: number;
    time: Date;
  }[];
};
```

### fetchCandles

Historical prices of expired futures can be retrieved with this end point but
make sure to specify start time and end time.
[Docs](https://docs.ftx.com/?javascript#get-historical-prices)

example:

```ts
import { fetchCandles } from "https://deno.land/x/ftx@$VERSION/mod.ts";
await fetchCandles({ marketName: "BTC/USD", resolution: 300 });
```

parameter:

| name       | type                                       | required | description          |
| ---------- | ------------------------------------------ | -------- | -------------------- |
| candleArgs | `FetchCandles`                             | *        | required candle args |
| options    | `{ startTime?: number, endTime?: number }` |          | candle options       |

```ts
type FetchCandles = {
  marketName: string;
  resolution: (15 | 60 | 300 | 900 | 3600 | 14400 | 86400) | ({} & number);
};
```

returns:

```ts
type CandlesResponse = {
  success: true;
  result: {
    open: number;
    high: number;
    low: number;
    close: number;
    startTime: Date;
    volume: number;
    time: number;
  }[];
};
```


CHANGELOG.md
# [1.0.0-beta.2](https://github.com/coinset/ftx/compare/v1.0.0-beta.1...v1.0.0-beta.2) (2022-01-14)


### Features

* **public:** add public api of `fetchCandles` function ([1e97665](https://github.com/coinset/ftx/commit/1e9766578da889330bdc1b69176c6c1dbec48c0a))

# 1.0.0-beta.1 (2022-01-14)


### Bug Fixes

* **public:** fix return types of `fetchOrderBook` ([eb9e2d1](https://github.com/coinset/ftx/commit/eb9e2d1d088416b1b360246d6229319e457f3457))


### Features

* **public:** add public api of `fetchOrderBook` function ([608731f](https://github.com/coinset/ftx/commit/608731f733c0b41fde753488126657c75a70fc00))
* **public:** add public api of `fetchTicker` function ([744ad88](https://github.com/coinset/ftx/commit/744ad88a201db5a2961781afefadf0b57e85d2ad))
* **public:** add public api of `fetchTrades` function ([5531325](https://github.com/coinset/ftx/commit/5531325165ba33d8dd8fbc5a440f16ef3cd7b8d7))


public/market_test.ts
import { expect, test } from "../dev_deps.ts";
import { fetchMarket } from "./market.ts";
import { dataEquality } from "./markets_test.ts";

test("fetchMarkets", async () => {
  await expect(fetchMarket("BTC/USD")).resolves.toEqual({
    success: true,
    result: dataEquality,
  });
});


public/trades_test.ts
import {
  any,
  anyArray,
  anyBoolean,
  anyNumber,
  anyOf,
  expect,
  test,
} from "../dev_deps.ts";
import { fetchTrades } from "./trades.ts";

test("fetchTrades", async () => {
  await expect(fetchTrades("BTC/USD")).resolves
    .toEqual({
      success: true,
      result: anyArray({
        id: anyNumber(),
        liquidation: anyBoolean(),
        price: anyNumber(),
        side: anyOf(["buy", "sell"]),
        size: anyNumber(),
        time: any(Date),
      }),
    });
});


public/candles_test.ts
import { any, anyArray, anyNumber, expect, test } from "../dev_deps.ts";
import { fetchCandles } from "./candles.ts";

test("fetchCandles", async () => {
  await expect(fetchCandles({ marketName: "BTC/USD", resolution: 300 }))
    .resolves.toEqual({
      success: true,
      result: anyArray({
        close: anyNumber(),
        high: anyNumber(),
        low: anyNumber(),
        open: anyNumber(),
        volume: anyNumber(),
        startTime: any(Date),
        time: anyNumber(),
      }),
    });
});


public/market.ts
import { BASE_URL } from "./constants.ts";
import { jsonFetch } from "./_utils.ts";

type BaseMarketData = {
  quoteVolume24h: number;

  /** Change in the past hour */
  change1h: number;

  /** Change in the past 24 hours */
  change24h: number;

  /** Change since start of day (00:00 UTC) */
  changeBod: number;

  highLeverageFeeExempt: boolean;

  /** Minimum maker order size (if >10 orders per hour fall below this size) */
  minProvideSize: number;

  enabled: boolean;

  /** best ask */
  ask: number | null;

  /** best bid */
  bid: number | null;

  /** last traded price */
  last: number | null;

  /** current price */
  price: number | null;

  /** if the market is in post-only mode (all orders get modified to be post-only, in addition to other settings they may have) */
  postOnly: boolean;

  priceIncrement: number;
  sizeIncrement: number;

  /** if the market has nonstandard restrictions on which jurisdictions can trade it */
  restricted: boolean;

  /** USD volume in past 24 hours */
  volumeUsd24h: number;
  tokenizedEquity?: boolean;
};

type MarketData = (
  & BaseMarketData
  & ({
    /** e.g. `"BTC/USD"` */
    name: `${string}/${string}`;
    baseCurrency: string;
    quoteCurrency: string;
    underlying: null;
    type: "spot";
  } | {
    name: string;
    baseCurrency: null;
    quoteCurrency: null;
    underlying: string;
    type: "future";
  })
);

export type MarketResponse = {
  success: true;
  result: MarketData;
};

/** Return type of market on FTX: spot, perpetual future, expiring future, and MOVE contract.
 * Examples for each type are `BTC/USD`, `BTC-PERP`, `BTC-0626`, and `BTC-MOVE-1005`. For future that expired in 2019, prepend a `2019` to the date, like so: `BTC-20190628` or `BTC-MOVE-20190923`.
 * ```ts
 * import { fetchMarket } from "https://deno.land/x/ftx@$VERSION/mod.ts"
 * await fetchMarket("BTC/USD")
 * ```
 * @see https://docs.ftx.com/?javascript#get-single-market
 */
export function fetchMarket(marketName: string): Promise<MarketResponse> {
  const url = new URL(`markets/${marketName}`, BASE_URL);

  return jsonFetch(url);
}

export type { MarketData };


public/trades.ts
import { BASE_URL } from "./constants.ts";
import { jsonFetch, Reviver } from "./_utils.ts";
import { isNumber, isString } from "../deps.ts";

export type TradesResponse = {
  success: true;
  result: {
    /** trade id */
    id: number;

    /** if this trade involved a liquidation order */
    liquidation: boolean;

    price: number;
    side: "buy" | "sell";
    size: number;
    time: Date;
  }[];
};

export type TradesOptions = {
  startTime?: number;
  endTime?: number;
};

const reviver: Reviver = (key, value) => {
  if (key === "time" && isString(value)) {
    return new Date(Date.parse(value));
  }
  return value;
};

/** Retrieve list of trade history.
 * ```ts
 * import { fetchTrades } from "https://deno.land/x/ftx@$VERSION/mod.ts";
 * await fetchTrades("BTC/USD");
 * ```
 * @see https://docs.ftx.com/?javascript#get-trades
 */
export function fetchTrades(
  marketName: string,
  { startTime, endTime }: TradesOptions = {},
  init?: RequestInit,
): Promise<TradesResponse> {
  const url = new URL(`markets/${marketName}/trades`, BASE_URL);

  if (isNumber(startTime)) {
    url.searchParams.set("start_time", String(startTime));
  }

  if (isNumber(endTime)) {
    url.searchParams.set("end_time", String(endTime));
  }

  return jsonFetch(url, init, {
    parseJson: reviver,
  });
}


public/markets.ts
import { BASE_URL } from "./constants.ts";
import { jsonFetch } from "./_utils.ts";
import { MarketData } from "./market.ts";

export type MarketsResponse = {
  success: true;
  result: MarketData[];
};

/** This section covers all types of markets on FTX: spot, perpetual futures, expiring futures, and MOVE contracts.
 * Examples for each type are `BTC/USD`, `BTC-PERP`, `BTC-0626`, and `BTC-MOVE-1005`. For futures that expired in 2019, prepend a `2019` to the date, like so: `BTC-20190628` or `BTC-MOVE-20190923`.
 * ```ts
 * import { fetchMarkets } from "https://deno.land/x/ftx@$VERSION/mod.ts"
 * await fetchMarkets()
 * ```
 * @see https://docs.ftx.com/?javascript#markets
 */
export function fetchMarkets(): Promise<MarketsResponse> {
  const url = new URL("markets", BASE_URL);

  return jsonFetch(url);
}


_tools/build_npm.ts
import { build } from "https://deno.land/x/dnt@0.7.3/mod.ts";

import {
  dirname,
  fromFileUrl,
  resolve,
} from "https://deno.land/std@0.114.0/path/mod.ts";

const baseDir = resolve(dirname(fromFileUrl(import.meta.url)), "..");

await build({
  entryPoints: [resolve(baseDir, "mod.ts")],
  outDir: resolve(baseDir, "npm"),
  test: false,
  compilerOptions: {
    sourceMap: true,
  },
  package: {
    name: "@coinset/ftx",
    version: Deno.args[0]?.replace(/^v/, ""),
    description: "Universal FTX API client",
    license: "MIT",
    sideEffects: false,
    "publishConfig": {
      "access": "public",
    },
    author: {
      name: "TomokiMiyauci",
      email: "contact@miyauchi.dev",
      url: "https://miyauchi.dev/",
    },
    repository: {
      type: "git",
      url: "https://github.com/coinset/ftx.git",
    },
    bugs: {
      url: "https://github.com/coinset/ftx/issues",
    },
    homepage: "https://github.com/coinset/ftx#readme",
    funding: {
      type: "patreon",
      url: "https://www.patreon.com/tomoki_miyauci",
    },
    keywords: [
      "ftx",
    ],
  },
});

Deno.copyFileSync("LICENSE", "npm/LICENSE");
Deno.copyFileSync("README.md", "npm/README.md");


public/order_book.ts
import { BASE_URL } from "./constants.ts";
import { jsonFetch } from "./_utils.ts";
import { isNumber } from "../deps.ts";

export type OrderBookResponse = {
  success: true;
  result: {
    /** Array with price and size */
    asks: [number, number][];

    /** Array with price and size */
    bids: [number, number][];
  };
};

export type OrderBookOptions = {
  /** Maximum is `100`
   * @default 20
   */
  depth?: number;
};

/** Retrieve order book of market name.
 * ```ts
 * import { fetchOrderBook } from "https://deno.land/x/ftx@$VERSION/mod.ts";
 * await fetchOrderBook("BTC/USD");
 * ```
 * @see https://docs.ftx.com/?javascript#get-orderbook
 */
export function fetchOrderBook(
  marketName: string,
  { depth }: OrderBookOptions = {},
): Promise<OrderBookResponse> {
  const url = new URL(`markets/${marketName}/orderbook`, BASE_URL);

  if (isNumber(depth)) {
    url.searchParams.set("depth", String(depth));
  }

  return jsonFetch(url);
}


public/candles.ts
import { BASE_URL } from "./constants.ts";
import { jsonFetch, Reviver } from "./_utils.ts";
import { isNumber, isString } from "../deps.ts";

export type CandlesResponse = {
  success: true;
  result: {
    /** mark price at startTime */
    open: number;

    /** highest mark price over the window */
    high: number;

    /** lowest mark price over the window */
    low: number;

    /** mark price at the end of the window: startTime + resolution */
    close: number;

    /** start time of the window */
    startTime: Date;

    /** volume traded in the window */
    volume: number;
    time: number;
  }[];
};

export type FetchCandles = {
  /** name of the market */
  marketName: string;

  /** window length in seconds. */
  // deno-lint-ignore ban-types
  resolution: (15 | 60 | 300 | 900 | 3600 | 14400 | 86400) | ({} & number);
};

export type CandleOptions = {
  /** filter starting time in seconds */
  startTime?: number;

  /** filter ending time in seconds */
  endTime?: number;
};

const reviver: Reviver = (key, value) => {
  if (key === "startTime" && isString(value)) {
    return new Date(Date.parse(value));
  }
  return value;
};

/** Historical prices of expired futures can be retrieved with this end point but make sure to specify start time and end time.
 * ```ts
 * import { fetchCandles } from "https://deno.land/x/ftx@$VERSION/mod.ts";
 * await fetchCandles({ marketName: "BTC/USD", resolution: 300 });
 * ```
 * @see https://docs.ftx.com/?javascript#get-historical-prices
 */
export function fetchCandles(
  { marketName, resolution }: FetchCandles,
  { startTime, endTime }: CandleOptions = {},
  init?: RequestInit,
): Promise<CandlesResponse> {
  const url = new URL(`markets/${marketName}/candles`, BASE_URL);

  url.searchParams.set("resolution", String(resolution));

  if (isNumber(startTime)) {
    url.searchParams.set("start_time", String(startTime));
  }

  if (isNumber(endTime)) {
    url.searchParams.set("end_time", String(endTime));
  }

  return jsonFetch(url, init, {
    parseJson: reviver,
  });
}
